Add UnsafeRawPointer type and API.

atrick opened the pull request:
As proposed in SE-0107:   UnsafeRawPointer.
https://github.com/apple/swift-evolution/blob/master/proposals/0107-unsaferawpointer.md

The fundamental difference between Unsafe[Mutable]RawPointer and
Unsafe[Mutable]Pointer<Pointee> is simply that the former is used for "untyped"
memory access, and the later is used for "typed" memory access. Let's refer to
these as "raw pointers" and "typed pointers". Because operations on raw pointers
access untyped memory, the compiler cannot make assumptions about the underlying
type of memory and must be conservative. With operations on typed pointers, the
compiler may make strict assumptions about the type of the underlying memory,
which allows more aggressive optimization.

Memory can only be accessed by a typed pointer when it is currently
bound to the Pointee type. Memory can be bound to type `T` via:
- `UnsafePointer<T>.allocate(capacity: n)`
- `UnsafePointer<Pointee>.withMemoryRebound(to: T.self, capacity: n) {...}`
- `UnsafeMutableRawPointer.initializeMemory(as: T.self, at: i, count: n, to: x)`
- `UnsafeMutableRawPointer.initializeMemory(as: T.self, from: p, count: n)`
- `UnsafeMutableRawPointer.moveInitializeMemory(as: T.self, from: p, count: n)`
- `UnsafeMutableRawPointer.bindMemory(to: T.self, capacity: n)`

Mangle UnsafeRawPointer as predefined substitution 'Sv' for Swift void
pointer ([urp] are taken).

atrick committed:
Add UnsafeRawPointer type and API.

As proposed in SE-0107:   UnsafeRawPointer.
https://github.com/apple/swift-evolution/blob/master/proposals/0107-unsaferawpointer.md

The fundamental difference between Unsafe[Mutable]RawPointer and
Unsafe[Mutable]Pointer<Pointee> is simply that the former is used for "untyped"
memory access, and the later is used for "typed" memory access. Let's refer to
these as "raw pointers" and "typed pointers". Because operations on raw pointers
access untyped memory, the compiler cannot make assumptions about the underlying
type of memory and must be conservative. With operations on typed pointers, the
compiler may make strict assumptions about the type of the underlying memory,
which allows more aggressive optimization.

Memory can only be accessed by a typed pointer when it is currently
bound to the Pointee type. Memory can be bound to type `T` via:
- `UnsafePointer<T>.allocate(capacity: n)`
- `UnsafePointer<Pointee>.withMemoryRebound(to: T.self, capacity: n) {...}`
- `UnsafeMutableRawPointer.initializeMemory(as: T.self, at: i, count: n, to: x)`
- `UnsafeMutableRawPointer.initializeMemory(as: T.self, from: p, count: n)`
- `UnsafeMutableRawPointer.moveInitializeMemory(as: T.self, from: p, count: n)`
- `UnsafeMutableRawPointer.bindMemory(to: T.self, capacity: n)`

Mangle UnsafeRawPointer as predefined substitution 'Sv' for Swift void
pointer ([urp] are taken).

atrick commented:
@swift-ci Please test and merge.

atrick commented:
@swift-ci Please test and merge.

gribozavr commented on test/1_stdlib/UnsafeRawPointer.swift:
```diff
@@ -0,0 +1,231 @@
+var UnsafeMutableRawPointerExtraTestSuite =
+  TestSuite("UnsafeMutableRawPointerExtra")
+
+class Missile {
```
Why not `LifetimeTracked`?  The harness will do leak checks for you.

atrick replied:
I made a note to follow up on this because I should also change the original UnsafePointer tests. For now the tests are consistent.

gribozavr commented on test/1_stdlib/UnsafeRawPointer.swift:
```diff
@@ -0,0 +1,231 @@
+  }
+}
+
+UnsafeMutableRawPointerExtraTestSuite.test("initializeMemory:as:from:count") {
```
`initializeMemory(as:from:count:)`

(and similarly in test names below)

atrick replied:
No problem :)

gribozavr commented on stdlib/public/core/UnsafeRawPointer.swift.gyb:
```diff
@@ -0,0 +1,557 @@
+      var destPtr = self + (count - 1)
+      var srcPtr = source + (count - 1)
+      while destPtr >= self {
+        Builtin.int_memcpy_RawPointer_RawPointer_Int64(
```
Could you wrap this builtin in a transparent function, just to add required labels and use non-builtin types?

atrick replied:
Yes I'll confine this imported builtin function and to Builtin.swift.

gribozavr commented on stdlib/public/core/UnsafeRawPointer.swift.gyb:
```diff
@@ -0,0 +1,557 @@
+  ///   different invocations of the same program.  Do not persist the
+  ///   hash value across program runs.
+  public var hashValue: Int {
+    return Int(Builtin.ptrtoint_Word(_rawValue))
```
`Int(bitPattern: self)`?

atrick replied:
That should work now. I'll change the original code that I copied.

gribozavr commented on test/1_stdlib/UnsafePointer.swift.gyb:
```diff
@@ -350,47 +396,16 @@ UnsafePointerTestSuite.test("customMirror") {
-  let isProperDisposition : Bool
+  let ptr = ${SelfType}(bitPattern: reallyBigInt)!
   if case let .text(desc) = ptr.customPlaygroundQuickLook {
 #if arch(i386) || arch(arm)
```
What was the issue with the tests above?

atrick replied:
No tests got deleted. Previously the tests were copy-pasted for no reason. Now one copy of the tests works for all pointer types, including raw pointers.

gribozavr replied:
I see, we have duplicates in `test/1_stdlib/UnsafePointer.swift.gyb` right now.

gribozavr commented on stdlib/public/core/UnsafeRawPointer.swift.gyb:
```diff
@@ -0,0 +1,557 @@
+}
+
+extension Unsafe${Mutable}RawPointer : CustomReflectable {
+  public var customMirror: Mirror {
```
Could you add tests for printing and reflection, like we have for `UnsafeMutablePointer`?

atrick replied:
Tests that are the same for UnsafeRawPointer and UnsafePointer are already inside UnsafePointer.switft.gyb. This file is only for tests that are unique to raw pointer.

gribozavr commented on stdlib/public/core/UnsafeRawPointer.swift.gyb:
```diff
@@ -0,0 +1,557 @@
+    let ptrValue = UInt64(
+      bitPattern: Int64(Int(Builtin.ptrtoint_Word(_rawValue))))
+    return ptrValue == 0
+    ? "\(selfType)(nil)"
```
I thought it can't store a null value.

atrick replied:
I'm maintaining consistency with UnsafePointer. I think we're keeping the (nil) printing for debugging purposes. @jrose-apple ?

jrose-apple replied:
That's why I left it in the old one, yes. But printing `0x0` is probably fine for debugging too.

gribozavr commented on stdlib/public/core/UnsafeRawPointer.swift.gyb:
```diff
@@ -0,0 +1,557 @@
+  }
+
+  public var customPlaygroundQuickLook: PlaygroundQuickLook {
+    return .text(summary)
```
Could you inline the calculation of summary into its only use site?

gribozavr replied:
Please consider doing the same in the file where this code was copied from.

atrick replied:
I wasn't sure why it was done that way. Maybe we want these things to have a summary property? I made a not to follow up in a separate PR.

atrick replied:
I have a "note" to follow up after finding out who's the expert here...

gribozavr replied:
I think it is unimportant historical reasons.  The `summary` property is internal, so it is not API.

gribozavr commented on lib/Basic/Demangle.cpp:
```diff
@@ -1037,6 +1037,10 @@ class Demangler {
       return createSwiftType(Node::Kind::Structure, "Float");
     if (Mangled.nextIf('i'))
       return createSwiftType(Node::Kind::Structure, "Int");
+    if (Mangled.nextIf('V'))
```
Tests for the mangler and demangler?

atrick replied:
Yes, I'll add those.

gribozavr commented on stdlib/public/core/UnsafePointer.swift.gyb:
```diff
@@ -336,18 +338,40 @@ public struct ${Self}<Pointee>
+  ///   is bound to `Pointee`.
+  public func withMemoryRebound<T, Result>(to: T.Type, capacity count: Int,
+    _ body: @noescape (UnsafeMutablePointer<T>) throws -> Result) rethrows
+  -> Result {
```
Could you format it in the usual stdlib style?

```swift
public func withMemoryRebound<T, Result>(
  to: T.Type,
  capacity count: Int,
  _ body: @noescape (UnsafeMutablePointer<T>) throws -> Result
) rethrows -> Result {
```

atrick replied:
Done.

gribozavr commented on stdlib/public/core/UnsafeRawPointer.swift.gyb:
```diff
@@ -0,0 +1,557 @@
+    _rawValue = Builtin.inttoptr_Word(bitPattern._builtinWordValue)
+  }
+
+  /// Convert an Unsafe${Mutable}Pointer ${a_Self}.
```
Convert ... *to*?

atrick replied:
I'll rewrite all the initializer comments to read "Convert <something> to ${a_Self}."

gribozavr commented on stdlib/public/core/UnsafeRawPointer.swift.gyb:
```diff
@@ -0,0 +1,557 @@
+    _rawValue = other._rawValue
+  }
+
+  /// Convert an Unsafe${Mutable}Pointer ${a_Self}.
```
Convert ... *to*?

gribozavr commented on stdlib/public/core/UnsafeRawPointer.swift.gyb:
```diff
@@ -0,0 +1,557 @@
+  }
+
+%  if not mutable:
+  /// Convert an UnsafeMutableRawPointer ${a_Self}.
```
Convert ... *to*?

gribozavr commented on stdlib/public/core/UnsafeRawPointer.swift.gyb:
```diff
@@ -0,0 +1,557 @@
+    _rawValue = other._rawValue
+  }
+
+  /// Convert an UnsafeMutableRawPointer ${a_Self}.
```
Convert ... *to*?

gribozavr commented on stdlib/public/core/UnsafeRawPointer.swift.gyb:
```diff
@@ -0,0 +1,557 @@
+    _rawValue = unwrapped._rawValue
+  }
+
+  /// Convert an UnsafeMutablePointer ${a_Self}.
```
Convert ... *to*?

gribozavr commented on stdlib/public/core/UnsafeRawPointer.swift.gyb:
```diff
@@ -0,0 +1,557 @@
+    _rawValue = other._rawValue
+  }
+
+  /// Convert an UnsafeMutablePointer ${a_Self}.
```
Convert ... *to*?

gribozavr commented on stdlib/public/core/UnsafeRawPointer.swift.gyb:
```diff
@@ -0,0 +1,557 @@
+  /// - Precondition: The memory is uninitialized.
+  ///
+  /// - Postcondition: The memory is bound to 'T' starting at `self` continuing
+  ///   through `self` + `capacity` * `strideof(T.self)`
```
Doc comments should refer to parameter names, so this line should use `count` instead of `capacity` (or you need to remove `count` from the function signature and keep the default name `capacity`).

atrick replied:
Yep, I just missed that one.

gribozavr commented on stdlib/public/core/UnsafeRawPointer.swift.gyb:
```diff
@@ -0,0 +1,557 @@
+  /// value of type `T`.
+  ///
+  /// - Precondition: The underlying pointer plus `offset` is properly
+  ///   aligned for accessing `T`.
```
Did you consider adding a debug precondition for alignment?  I expect this to be frequently violated.

gribozavr replied:
Similarly for `storeBytes()`.

atrick replied:
I did consider it and forgot to do it. Will do.

gribozavr commented on stdlib/public/core/UnsafeRawPointer.swift.gyb:
```diff
@@ -0,0 +1,557 @@
+      bitPattern: Int64(Int(Builtin.ptrtoint_Word(_rawValue))))
+    return ptrValue == 0
+    ? "\(selfType)(nil)"
+    : "\(selfType)(0x\(_uint64ToString(ptrValue, radix:16, uppercase:true)))"
```
You could use `_rawPointerToString()`.

atrick replied:
The formatting is a bit different. Upper/lower case, padding with zeros. I'm not really the expert on this UI.

atrick commented:
Assuming this merge goes through, I'll create a new PR with a slew of test case tweaks, comments and formatting based on Dmitri's comments.

atrick committed:
UnsafeRawPointer minor improvements.

Incorporate Dmitri's feedback.

Properly use a _memmove helper.

Add load/storeBytes alignment precondition checks.

Reword comments.

Demangler tests.

atrick commented:
Don't merge this yet. UnsafePointer mangling test cases don't work for some reason.

atrick committed:
Fix name mangling test cases.

atrick commented:
@swift-ci Please test and merge.

atrick commented:
This depends on a generic substitution feature that need to pull in first.

atrick committed:
Fix bind_memory specialization.

atrick commented:
This passes the "long" tests now.
@swift-ci Please smoke test and merge.

atrick commented:
This symbol is missing from corelibs-foundation:
Swift.== infix (Swift.String, Swift.String) -> Swift.Bool
